parcels/particleset/baseparticleset.py

```diff
@@ -433,12 +446,34 @@ def execute(self, pyfunc=AdvectionRK4, endtime=None, runtime=None, dt=1.,
                                 'to a NetCDF file during the run.' % output_file.tempwritedir_base)
                 pbar = self.__create_progressbar(_starttime, endtime)
                 verbose_progress = True
+
             if dt > 0:
-                time = min(next_prelease, next_input, next_output, next_movie, next_callback, endtime)
+                next_time = min(next_prelease, next_input, next_output, next_movie, next_callback, endtime)
+            else:
+                next_time = max(next_prelease, next_input, next_output, next_movie, next_callback, endtime)
+
+            # qubixes: I am not sure if this is how it is supposed to work..
+            if self.interaction_kernel is None:
+                self.kernel.execute(self, endtime=next_time, dt=dt, recovery=recovery, output_file=output_file,
+                                    execute_once=execute_once)
             else:
-                time = max(next_prelease, next_input, next_output, next_movie, next_callback, endtime)
-            self.kernel.execute(self, endtime=time, dt=dt, recovery=recovery, output_file=output_file,
-                                execute_once=execute_once)
+                cur_time = time
+                while (cur_time < next_time and dt > 0) or (cur_time > next_time and dt < 0) or dt == 0:
+                    if dt > 0:
+                        cur_end_time = min(cur_time+dt, next_time)
+                    else:
+                        cur_end_time = max(cur_time+dt, next_time)
+                    self.kernel.execute(
+                        self, endtime=cur_end_time, dt=dt, recovery=recovery,
+                        output_file=output_file, execute_once=execute_once)
+
+                    self.interaction_kernel.execute(
+                        self, endtime=cur_end_time, dt=dt, recovery=recovery,
+                        output_file=output_file, execute_once=execute_once)
+                    cur_time += dt
+                    if dt == 0:
+                        break
+            time = next_time
```

Comment:
I object to putting that in the `BaseParticleSet`, as it performs those calculation-internal loops very slowly in Python, instead of doing that fast in C if the kernel is 'jit' and one doesn't do interaction. It kills the performance one gains from the C-compiled section. That is because here you basically require than the jit kernel to return and return to this outside loop, even though that may not be necessary.

I also wouldn't feel comfortable doing that in the `ParticleSetSOA::execiute(...)` out of the reason above and because the `kernel.execute(...)` should not be looped here. It is exactly because `cur_time` is not just that min- or max statement, as the field input advances asyncronously to it. If it was that simple, we'd done that months ago. I'd be curious about the test results of `test_smalldt(...)` for `'soa'` and `'aos'` for this code section, because that would report the break.

Reply:
I think it is because of that why I initially in our class refactoring objected to put the `execute(...)` function into the base class - the function is too specific for each structure and each feature to change it without affecting other structures or features. But yeah: I'll look into the tests and this also needs a benchmark to see if that slows down the non-interaction SoA processing.

Reply:
Purely regarding the inheritance-issue (leaving the time-thing out of this argument for now) it is a bit of a dilemma.

As it is currently implemented, the code in the `execute` function of the `BaseParticleSet` only differs in these lines. The rest of the execute function is unchanged. Creating an entire `execute_interaction` function would mean a lot of code duplication.

Alternatively, you could define a function that only contains the bit that is in the else-clause now (i.e. executed if an interactionkernel is defined), and call that function. This function could then be implemented in the subclass. But I do not see the benefit of moving that into the subclass, as the code here does not seem datastructure-specific to me, and should work for any (future) datastructure that would implement InteractionKernels.

In general, as long as the general structure of the execute function is more or less preserved across datastructures, I would favor keeping the implementation in the baseclass as a templating method, that refers to other functions (that can be overriden in subclasses) for datastructure-specific parts of the code (such as `_set_particle_vector` in the current execute function). But you have a better overview of the difficulties this might bring, through your experience implementing AoS, node list etc., so I will defer to your final judgement on this.

I would not expect significant changes in the performance of runs without InteractionKernels, as the "new" code is not touched in that case, at the cost of one extra if-statement. Kernels cannot be JIT if you do interaction. Or am I missing something there?

Reply:
Let me reply in pleces, please.
a) 
> As it is currently implemented, the code in the execute function of the `BaseParticleSet` only differs in these lines. 
> The rest of the execute function is unchanged. Creating an entire `execute_interaction` function would mean 
> a lot of code duplication.

I agree on the part that it would be lots of code duplication. Perhaps it would be good in the aftermath to look at that execute-function and split it into more logical atomic functions that can be easier handled in a modular manner.

> But I do not see the benefit of moving that into the subclass, as the code here does not seem datastructure-specific to 
> me, and should work for any (future) datastructure that would implement InteractionKernels.

Generally I'd agree, but it is exactly that I doubt this to work for general applications in an *equal* manner as the code before, *equal* referring to (a) floating-point accuracy and (b) JIT performance. And because they are not equal, I think it's presumptuous to extend a constraint for a limited-scope functionality to all simulations of the framework.

> In general, as long as the general structure of the execute function is _more or less preserved_ across datastructures

Again, if all functions and code structures were *equal*, achieving performance and provable working would not be an issue. Obviously, that is not the case. 

> I would not expect significant changes in the performance of runs without InteractionKernels, as the "new" code is not 
> touched in that case, at the cost of one extra if-statement. Kernels cannot be JIT if you do interaction. Or am I missing 
> something there?

Well, that's the bear in the closet. Look, before this, the timing issue was handled in the Kernel, One of the reasons (apart from keeping the JIT and Python Kernel look as much the same as possible to lend non-C developers a hand when fixing JIT ctype/cgen issues) is that in the Python as well as the JIT-Kernel, we check the timing and progress `dt` - as long as it is within `endtime`. Thus, the kernel steers the time, because it is also the Kernel that requests new data if required, it's the kernel that starts delayed particles, it's the Kernel that deletes particles and so forth. That means that e.g. the (JIT and Python) Kernel only returns when data needs to be put out, be it in file or commandline or in a plot, if for evaluating signals (e.g. recovery).
With the adaptation, the whole time-progression and so on is done in the Python main function. Now, that means that even without the interaction component, the AoS and SoA kernels (which both use that `BaseParticleSet::execute(...)`) will require the kernels now to return for each `dt` (and not just each output) to Python-mode. Above that, now even worse, it seems the time-progression thing for SoA/AoS-JIT is now done double - once in the Kernel and once in the ParticleSet execute. 

That is not how that should be. If you can demo me with the smalldt, long time-run simulations that this is not the case or not an issue, I may be swayed here - perhaps run simulations from the people who ran into the issue, such as @cjongedijk or @reint-fischer . They would probably be really sad if suddenly all their work on getting the time integration stable is suddenly reverted - just saying that this time integration, to get it as stable as it is today, we spend overall around 2-3 months within the last 1.5 years on that one thing. We tried out many simplified as well as complex versions, many ideas from many people, lots of them looked good in the beginning but then failed real-world inspection, re-opening the issue time and time again. I don't say "this is the way" and the only way this function to work, but if making time progression 'kind-a work' for this one interaction Kernel requires that big of a change, then I think it's better to take that issue separately than causing more trouble than it's worth. I still like your contribution, which is why I suggested to just move it and make it's execution conditional to the scope it is meant to cover, instead of just reverting it entirely, and it may be a good starting point for later simplification. Also, just to say: the *easiest* way by far to do integration would be to not progress `time` and check `time`, but to loop over integer-integration steps and check if there is some 'special action' between the current and the next timestep (i.e. *integer integration*). That's how it's done in many time-stable simulations of fluids in CS (e.g. Physics engines in games) - sadly enough, that we can't do without basically rewriting the whole core. So, yes, there are simpler ways to write simulation main loops, but for the many constraints of the many sub-functionalities in Parcels, they just don't apply to the system as it is by now.

Now, in the end, I agree that code duplication isn't nice, and we did a lot with the subclassing already to avoid it. In this case though, this prototype-version does a major change to the main processing function which has implications outside of its scope. In order to prevent collisions, there are, I think, two options: (a) allow code duplication, at least for a little while, change as proposed, and we can modularise the execute function soon together - that shouldn't be too difficult or take too long, but we just take it separate from this interaction issue; (b) we modularise now - we can do it, but then I think even when resolving the other open requests, it may delay the wrap-up of this issue significantly. I'd opt for (a), as we will need to create some follow-up issues (or: TODO's) from this PR anyway, so it'll be just one more item on the bucketlist.

Reply:
>With the adaptation, the whole time-progression and so on is done in the Python main function. Now, that means that even without the interaction component, the AoS and SoA kernels (which both use that BaseParticleSet::execute(...)) will require the kernels now to return for each dt (and not just each output) to Python-mode. Above that, now even worse, it seems the time-progression thing for SoA/AoS-JIT is now done double - once in the Kernel and once in the ParticleSet execute.

That's not true; with no interaction (`self.interaction_kernel is None`) all time-progression is done in the same way at is currently done in the main branch of parcels. The performance implications for the alteration for non-interacting kernels should be one if-statement per `outputdt` (not `dt`). Given the code around it, I doubt it would have any noticeable performance impact. If we/I still don't understand each other, then perhaps a teams call is more efficient in figuring this out?

Related to this part of the code, I wrote a small test to check whether time progression is implemented in the same way with or without interaction. It took me a while to understand what is going on under the hood, but finally it passes the test. I did however change the order of the interaction/non-interaction kernels; I hope that's not really an issue. The new test is at the end of `test_interaction_kernel.py`

Reply:
I'll have a look on that. Thanks for the feedback.

Reply:
see new comment on benchmarks.